Promote GlobalAccess for GCE Internal Loadbalancers to GA

skmatti opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
 /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:
Promotes Global Access for GCE Internal Loabalancers to use GA API

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #88390

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

skmatti committed:
Use GA API for ILB Global Access

skmatti commented:
/assign @MrHohn

skmatti commented:
/assign @prameshj

skmatti commented:
/kind cleanup

MrHohn commented:
/ok-to-test

MrHohn reviewed (approved):
/lgtm
/approve

prameshj commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -900,15 +899,16 @@ func toForwardingRuleComposite(rule interface{}) (frc *forwardingRuleComposite,
 	switch fr := rule.(type) {
 	case *compute.ForwardingRule:
 		frc = &forwardingRuleComposite{
```
the forwardingRuleComposite struct was added to distinguish between Beta and GA api. Now that we are calling GA api whether or not GlobalAccess is enabled, we can get rid of this struct and use compute.ForwardingRule directly.

fedebongio commented:
/remove-sig api-machinery

skmatti commented:
etcd references a few deprecated fields of grpc 1.26.0. Pinning grpc to 1.27.0 breaks etcd as grpc removed backward compatibility for those fields. https://github.com/etcd-io/etcd/issues/11563
A fix is on the way: https://github.com/etcd-io/etcd/pull/11564

skmatti committed:
Update dependency google.golang.org/api to v0.15.1

Steps:
./hack/pin-dependency.sh google.golang.org/api v0.15.1
./hack/update-vendor.sh
./hack/verify-vendor.sh
./hack/pin-dependency.sh github.com/googleapis/gax-go/v2 v2.0.5
./hack/update-vendor.sh
./hack/verify-vendor.sh
./hack/lint-dependencies.sh

skmatti commented:
/retest

skmatti commented:
/retest

skmatti commented:
/test pull-kubernetes-integration

skmatti commented:
/retest

prameshj commented:
Need to run `./hack/update-gofmt.sh` for pull-kubernetes-verify to pass.

/priority important-soon

skmatti commented:
/retest

skmatti commented:
/test pull-kubernetes-e2e-kind

skmatti commented:
/retest

skmatti commented:
/retest

MrHohn commented on staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_internal.go:
```diff
@@ -1056,23 +927,20 @@ func (g *Cloud) ensureInternalForwardingRule(existingFwdRule *compute.Forwarding
+}
+
+func forwardingRulesEqual(old, new *compute.ForwardingRule) bool {
+	return (old.IPAddress == "" || new.IPAddress == "" || old.IPAddress == new.IPAddress) &&
```
Trying to understand the logic - isn't `old.IPAddress == new.IPAddress` sufficient?

skmatti replied:
This logic was introduced in #46663 (Search for `FwdRuleEqual` in `gce_loadbalancer_internal.go`)

@prameshj brought this up as well, we are still figuring out the thought process behind this logic. Probably, I will fix this in a follow up PR.

prameshj replied:
It seems fine to replace with just "old.IPAddress == new.IPAddress". I am thinking the `new.IPAddress == ""` is to disallow change when IP assignment drops to empty if network/subnetwork have not changed.

I can't see why `old.IPAddress== ""` needs to be here. We were thinking of fixing in a follow-up PR so that in can be easily reverted without reverting the GA api change.

MrHohn reviewed (approved):
Thanks, LGTM with a minor question.

MrHohn commented:
/lgtm
/approve

prameshj commented:
/lgtm

skmatti commented:
/assign @wojtek-t

MrHohn commented:
/lgtm

wojtek-t commented:
/approve

skmatti commented:
/retest

skmatti commented:
/retest

skmatti commented:
/test pull-kubernetes-e2e-kind
